### ISSUE TITLE
Keycloak service is broken since upgrade to 15.0

### ISSUE THREAD

**Issue by dhess** [NixOS/nixpkgs#135313]
### Describe the bug

Ever since #132225 (Keycloak upgrade to 15.0), our Keycloak installation is broken. The admin console doesn't work (admin password is incorrect) and realms have disappeared. The service appears to be running just fine, but the configuration seems to be missing.

Downgrading to 14.0 and redeploying restores Keycloak to a working state again.

We've changed no NixOS configuration from one version to the next, this is just a `nixpkgs` bump.

### Steps To Reproduce

1. Working Keycloak 14 installation with realms etc.
2. Bump `nixpkgs-unstable` to latest.
3. Re-deploy keycloak service.

### Notify maintainers

@ngerstle @talyz

**Comment by mohe2015** [comment]
You could try a git bisect if you want to put in the effort.

**Comment by dhess** [comment]
Yes, I'm aware, but the prospect of running a `git bisect` on a production IdP that forgets its database on each bad deploy is not really practical. I was hoping that by posting this, somebody might have an idea of what could have gone wrong.

**Comment by mohe2015** [comment]
Which database are you using? Did you check your system logs?

**Comment by mohe2015** [comment]
If postgresql maybe https://github.com/NixOS/nixpkgs/pull/ 131018

**Comment by talyz** [comment]
Does indeed sound like the connection to the database was lost and it reverted to using the embedded one. I'll see if I can reproduce it and if the default module configuration has to be updated.

**Comment by dhess** [comment]
> Does indeed sound like the connection to the database was lost and it reverted to using the embedded one. I'll see if I can reproduce it and if the default module configuration has to be updated.

Ahh, that sounds like a very plausible explanation for what's going on. When the service finally starts, it comes up with no obvious errors, but it's so verbose, I didn't notice whether it mentioned the database location.

I'll take a look at the logs more carefully when I have a chance.

**Comment by dhess** [comment]
Here are the logs from the upgrade attempt, from the point where the previous service went down:

https://gist.github.com/dhess/8a50e967a0d6a43f36acbc05f27cbc36

It does appear that some database operations failed, though I confess I understand neither Keycloak nor JBoss/Wildfly well enough to understand what, exactly, went wrong.

I'm fairly certain that previous upgrades of our Keycloak service have had similar PostgreSQL failures, but in the past the service eventually always comes up working, just not this time.

**Comment by talyz** [comment]
I've tried to reproduce this issue without any "luck". I started with the first commit you referenced, imported a copy of my production database, upgraded to the second commit and then to the latest master and everything just worked as expected.

The error in your log seems to be due to Keycloak being forced to quit early in its boot process, for some reason. The subsequent start seems to be just fine, though. I can see that you're using PostgreSQL 13.3. If you previously used an older version, did you manually migrate to 13? Just switching to a newer version of PostgreSQL, without migrating the data, would give you an empty database, so that's why I'm asking.

**Comment by dhess** [comment]
I've never done a PostgreSQL migration at all. I've just been bumping nixpkgs versions and everything has worked until now. All settings, realms, etc. have worked just fine all along.

**Comment by dhess** [comment]
I was considering doing a migration from a local PostgreSQL instance to RDS Aurora (the PostgreSQL-compatible version). Maybe now's the time to try that with a new Keycloak instance running the new version.

**Comment by mohe2015** [comment]
> I've never done a PostgreSQL migration at all. I've just been bumping nixpkgs versions and everything has worked until now. All settings, realms, etc. have worked just fine all along.

Which PostgreSQL package are you using? Because if you use "pkgs.postgresql" and didn't set your stateVersion you are not following the manuals as you need to follow the release notes for every release. This leads to a nonupgraded database but an upgraded database server and therefore an empty database

**Comment by dhess** [comment]
I'm not overriding the PostgreSQL package in my Keycloak config, nor anywhere else in the system's config. Neither am I explicitly setting `stateVersion`. What `stateVersion` should I be setting on current `nixpkgs-unstable`?

**Comment by talyz** [comment]
Okay, normally, PostgreSQL would be kept at the same major version until you manually upgrade or bump the `system.stateVersion`, but if it's not set at all, it will just default to the latest (`21.11`, currently).

**Comment by talyz** [comment]
You should have it set to what was the default when you initially deployed the machine (probably `21.05` in your case). That will lock down default values that needs manual intervention when upgrading.

**Comment by mohe2015** [comment]
This shows that https://github.com/NixOS/nixpkgs/pull/124261 is really needed :-)

**Comment by mohe2015** [comment]
@dhess
If you want to use unstable somewhat stabley as far as I understand it you should set `stateVersion` to the latest released version (`21.05` currently). This is general advice.

This should revert PostgreSQL back to version 11 (https://github.com/NixOS/nixpkgs/blob/6248814b6892af7dc0cf973b49690fd102088e02/nixos/modules/services/databases/postgresql.nix#L292) which **should** have you existing data.

**Comment by mohe2015** [comment]
Alternatively in this custom case you could also manually set `services.postgresql.package = pkgs.postgresql_11` e.g. if you would like to set your `stateVersion` to `21.11` for some reason. As fas as I understand this is strongly discouraged for production nixos-unstable systems as it immediately applies breaking changes that are commited.

**Comment by dhess** [comment]
Thanks, @mohe2015, I'll try one of those suggestions and update this issue accordingly.

**Comment by talyz** [comment]
Yes, what @mohe2015 said. If you want to upgrade to the latest PostgreSQL version, however, it's not that hard either. Here is the script I use to perform the upgrade:
```nix
{
  environment.systemPackages =
    let
      mkPostgresUpgradeScript = oldPackage:
        (pkgs.writeScriptBin "upgrade-pg-cluster" ''
          set -eux
          export OLDDATA="/var/lib/postgresql/${oldPackage.psqlSchema}"
          export NEWDATA="${config.services.postgresql.dataDir}"
          export OLDBIN="${oldPackage}/bin"
          export NEWBIN="${config.services.postgresql.package}/bin"

          # Stop the new postgres server and remove all its
          # data. Services may already have created fresh users
          # and databases, which we don't want, since they would
          # conflict with the existing users and databases we're
          # migrating.
          systemctl stop
          postgresql rm -rf "$NEWDATA"

          # Initialize the new database
          sudo -u postgres initdb -U postgres -D "$NEWDATA" ${lib.concatStringsSep " " config.services.postgresql.initdbArgs}

          # Perform the upgrade
          cd $NEWDATA
          sudo -u postgres $NEWBIN/pg_upgrade \
            --old-datadir "$OLDDATA" --new-datadir "$NEWDATA" \
            --old-bindir $OLDBIN --new-bindir $NEWBIN \
            "$@"

          systemctl start postgresql

          sudo -u postgres "$NEWDATA/analyze_new_cluster.sh"
        '');
    in
      [
        (mkPostgresUpgradeScript pkgs.postgresql_11)
      ];
}
```
